### PR TITLE
add: DFlash block diffusion speculative decoding

### PR DESCRIPTION
**Pull request by ChenhanYu**
NOTE: closing this PR and opt for https://github.com/NVIDIA/Model-Optimizer/pull/1211

Implement DFlash (Block Diffusion for Flash Speculative Decoding) as a new mode in ModelOpt's speculative decoding framework.

Key architecture:
- Feature Fusion: extract hidden states from uniformly sampled target model layers, project via FC layer
- KV Injection: fused target features injected as K/V entries in every draft decoder layer's attention (not just first layer input)
- Parallel Drafting: all tokens in a block predicted simultaneously using learnable mask embeddings and bidirectional within-block attention

Files:
- dflash/ module: DFlashModel, DFlashConfig, conversion, default config
- plugins/hf_dflash.py: HFDFlashModel with DFlashAttention (KV injection), DFlashModule (feature fusion + decoder), training forward pass with random anchor sampling and exponential position decay loss
- main.py: --mode dflash support in training script

Reference: "DFlash: Block Diffusion for Flash Speculative Decoding" (arXiv:2602.06036)

### What does this PR do?

Type of change: ? 

### Usage

```python
# Add a code snippet demonstrating how to use this
```

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ / ❌ / N/A 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅ / ❌ / N/A 
- Did you write any new necessary tests?: ✅ / ❌ / N/A 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ / ❌ / N/A 

## Summary by CodeRabbit

* **New Features**
  * Added DFlash speculative-decoding mode with end-user config options and sensible defaults (block size, decoder layers, architecture JSON, torch.compile toggle, mask token).
  * CLI/training accepts and runs in DFlash mode; dataset flow supports DFlash.
  * HuggingFace DFlash plugin with draft-module training, draft forward, and pseudo‑speculative generation helpers.

* **Bug Fixes**
  * Validation/logging during training gated to rank‑0 with proper synchronization.

### CODE REVIEW

**Comment by h-guo18** [comment]
LGTM. Shall we also add some unit/example tests in this PR

File: modelopt/torch/speculative/plugins/hf_dflash.py [hunk 1]
```diff
@@ -0,0 +1,882 @@
...
+            if isinstance(module, nn.Linear):
+                nn.init.normal_(module.weight, mean=0.0, std=std)
+                if module.bias is not None:
+                    nn.init.zeros_(module.bias)
+
+    def forward(self, noise_embedding, target_hidden, position_ids, attention_mask=None):
+        """Forward matching SpecForge DFlashDraftModel.forward."""
+        hidden_states = noise_embedding
+        target_hidden = self.hidden_norm(self.fc(target_hidden))
+        position_embeddings = self.rotary_emb(hidden_states, position_ids)
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/hf_dflash.py:290]:
```suggestion
        #lazy init rope
        if not hasattr(self, "rotary_emb"):
           self.rotary_emb = _ROTARY_CLS(config=config, device=hidden_states.device)
        position_embeddings = self.rotary_emb(hidden_states, position_ids)
```

File: modelopt/torch/speculative/plugins/hf_dflash.py [hunk 2]
```diff
@@ -0,0 +1,882 @@
...
+        num_fused_layers = len(config.target_layer_ids)
+        self.fc = nn.Linear(num_fused_layers * config.hidden_size, config.hidden_size, bias=False)
+        self.hidden_norm = _NORM_CLS(config.hidden_size, eps=config.rms_norm_eps)
+
+        # Decoder layers
+        self.layers = nn.ModuleList(
+            [DFlashDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
+        )
+        self.norm = _NORM_CLS(config.hidden_size, eps=config.rms_norm_eps)
+        self.rotary_emb = _ROTARY_CLS(config=config)
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/hf_dflash.py:267]:
```suggestion
```

Lazy init as below to avoid meta tensor copy error

File: modelopt/torch/speculative/plugins/hf_dflash.py [hunk 3]
```diff
@@ -0,0 +1,882 @@
...
+    # Handle multi-word model types (e.g., "qwen3" -> "Qwen3")
+    for attr in dir(mod):
+        if attr.lower() == f"{model_type}mlp":
+            prefix = attr.replace("MLP", "")
+            break
+
+    mlp_cls = getattr(mod, f"{prefix}MLP", None)
+    norm_cls = getattr(mod, f"{prefix}RMSNorm", None)
+    rotary_cls = getattr(mod, f"{prefix}RotaryEmbedding", None)
+    rotate_half_fn = getattr(mod, "rotate_half", None)
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/hf_dflash.py:64]: Is there any reason we prefer matching target's component classes rather than fixing them to llama-components:?
